tripal/tripal
# Update Tripal Layout  to support configuring the Form Display

# Tripal 4 Core Dev Task

### Issue #1863 


### Tripal Version: 4

## Description


Tripal Layout currently allows module developers to define a layout of fields and fieldsets for the page display. We would like to expand the same functionality to handle the form display. For specific use cases, see the linked issue.

This PR creates a new config entity (TripalLayoutDefaultForm) to handle form display layouts based on the original TripalLayoutDefaultView config entity which supports configuring the default page display.

**NOTE: This PR includes an unavoidable BREAKING CHANGE.** All layout yamls made before this PR will need to be updated to match the new format. All core yamls have been updated. This was needed because the previous yaml was not supportable by Drupal translation. This was only caught when automated testing was added.

### Tasks
- [x] Create TripalLayoutDefaultForm config entity based on TripalLayoutDefaultView
- [x] Create the YAML to organize the form for one content type (i.e. organism) and ensure the config schema is in place to support that.
- [x] Ensure that new TripalLayoutDefaultForm entities are created for each YAML file matching the file naming pattern just like what happens for the display.
- [x] Add "Apply Tripal Layout" and "Reset Layout" buttons to the form display page.
- [x] Implement the functionality within the controller for the "Reset Layout" form display button.
- [x] Implement the functionality within the controller for the "Apply Tripal Layout" form display button.
- [x] Cleanup the controller to reduce code duplication as much as possible.
- [x] Fix schema errors caught by automated testing
- [x] Ensure child fields are in the order they are listed in the YAML.
- [x] Support descriptions for field groups
- [x] Remove any empty field groups
- [x] A form template for analysis content type.
- [x] Some/Any automated testing

## Testing?




### Automated Testing

This PR adds the first automated testing for the Tripal layout module. This is the coverage it adds:

<img width="1387" alt="Screenshot 2024-09-28 at 8 16 02 PM" src="https://github.com/user-attachments/assets/53f93220-8a1c-4d5a-ac53-e51738df20f2">

#### Testing still missing

 - controller: actually applies the layouts

### Manual Testing

On a fresh docker on this branch,
1. Go to Admin > Tripal > Configuration > Tripal Default Form Layouts (admin/tripal/config/tripal_layout_default_form) and confirm that there is an entry listed for "General Form".
2. Go to the Manage form display for the analysis content type: Tripal > Page Structure > Analysis > Manage Form Display (admin/structure/bio_data/manage/analysis/form-display) and confirm that there are "Apply Tripal Layout" and "Reset Layout" buttons added.
  ![image](https://github.com/user-attachments/assets/dcc2447e-b179-4146-9edc-53d0e0fb0c1c)
3. Make some manual changes to the form display and save. Then click the reset button and confirm they are reset.
4. Click the "Apply Tripal Layout" button and confirm the form is separated into fieldsets as designed.
  ![image](https://github.com/user-attachments/assets/05400e39-fab9-41a5-978c-ca83304ca3b0)
5. Now redo steps 2 and 3 for the "Manage Display". You should expect to see the following.
  ![image](https://github.com/user-attachments/assets/ec8c19a1-dc73-4c86-9b61-56a4c84478bc)
6. Create an analysis and confirm that the form shows this new form display layout and that no errors are encountered during save. This form will also be available on edit.
  ![image](https://github.com/user-attachments/assets/bcf58c27-b6b1-4453-b9cc-fedc96bd8c88)
7. After saving your analysis confirm the display matches the set template for "Manage Display". For example, like this.
  ![image](https://github.com/user-attachments/assets/4f55b9bf-f637-47ad-a562-bf45adfd4019)


Other things to test:
- confirm you can still customize the form display after the Tripal Layout has been applied and see these customization in the add/edit content form for that content type.
- confirm you can reset the tripal layout back to the drupal one.
- check the organism form display as well.
- try to apply a template for any of the other content types and expect to be told they don't have one.
- also test display layouts since the schema needed to change.
